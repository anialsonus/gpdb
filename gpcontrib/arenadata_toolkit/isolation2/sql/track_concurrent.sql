-- start_matchsubs
-- m/ERROR:  Track for database \d+ is being acquired in other transaction/
-- s/\d+/XXX/g
-- end_matchsubs
-- Test concurrent track acquisition.
1: CREATE EXTENSION IF NOT EXISTS arenadata_toolkit;
1: SELECT arenadata_toolkit.tracking_register_db();
1: SELECT arenadata_toolkit.tracking_trigger_initial_snapshot();
1: BEGIN;
1: WITH segment_counts AS (
    SELECT tt.segid, COUNT(*) AS cnt 
    FROM arenadata_toolkit.tables_track tt 
    GROUP BY tt.segid
),
pg_class_count AS (
    SELECT COUNT(*) AS cnt FROM pg_class c JOIN pg_namespace n ON c.relnamespace = n.oid
    WHERE nspname = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_schemas'), ','))
    AND c.relstorage = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_relstorages'), ','))
    AND c.relkind = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_relkinds'), ','))
)
SELECT bool_and(sc.cnt = pc.cnt)
FROM segment_counts sc, pg_class_count pc;

2: SELECT tt.segid, count(*) FROM arenadata_toolkit.tables_track tt GROUP BY tt.segid;

1: ROLLBACK;

2: WITH segment_counts AS (
    SELECT tt.segid, COUNT(*) AS cnt 
    FROM arenadata_toolkit.tables_track tt 
    GROUP BY tt.segid
),
pg_class_count AS (
    SELECT COUNT(*) AS cnt FROM pg_class c JOIN pg_namespace n ON c.relnamespace = n.oid
    WHERE nspname = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_schemas'), ','))
    AND c.relstorage = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_relstorages'), ','))
    AND c.relkind = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_relkinds'), ','))
)
SELECT bool_and(sc.cnt = pc.cnt)
FROM segment_counts sc, pg_class_count pc;

1: SELECT arenadata_toolkit.tracking_unregister_db();
1: DROP EXTENSION arenadata_toolkit;

1q:
2q:

!\retcode gpconfig -r shared_preload_libraries;
!\retcode gpconfig -r arenadata_toolkit.tracking_worker_naptime_sec;
!\retcode gpstop -raq -M fast;
