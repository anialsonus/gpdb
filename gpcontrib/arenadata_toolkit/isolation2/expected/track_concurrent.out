-- start_matchsubs
-- m/ERROR:  Track for database \d+ is being acquired in other transaction/
-- s/\d+/XXX/g
-- end_matchsubs
-- Test concurrent track acquisition.
1: CREATE EXTENSION IF NOT EXISTS arenadata_toolkit;
CREATE
1: SELECT arenadata_toolkit.tracking_register_db();
 tracking_register_db 
----------------------
 t                    
(1 row)
1: SELECT arenadata_toolkit.tracking_trigger_initial_snapshot();
 tracking_trigger_initial_snapshot 
-----------------------------------
 t                                 
(1 row)
1: BEGIN;
BEGIN
1: WITH segment_counts AS ( SELECT tt.segid, COUNT(*) AS cnt FROM arenadata_toolkit.tables_track tt GROUP BY tt.segid ), pg_class_count AS ( SELECT COUNT(*) AS cnt FROM pg_class c JOIN pg_namespace n ON c.relnamespace = n.oid WHERE nspname = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_schemas'), ',')) AND c.relstorage = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_relstorages'), ',')) AND c.relkind = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_relkinds'), ',')) ) SELECT bool_and(sc.cnt = pc.cnt) FROM segment_counts sc, pg_class_count pc;
 bool_and 
----------
 t        
(1 row)

2: SELECT tt.segid, count(*) FROM arenadata_toolkit.tables_track tt GROUP BY tt.segid;
ERROR:  Track for database 24798 is being acquired in other transaction

1: ROLLBACK;
ROLLBACK

2: WITH segment_counts AS ( SELECT tt.segid, COUNT(*) AS cnt FROM arenadata_toolkit.tables_track tt GROUP BY tt.segid ), pg_class_count AS ( SELECT COUNT(*) AS cnt FROM pg_class c JOIN pg_namespace n ON c.relnamespace = n.oid WHERE nspname = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_schemas'), ',')) AND c.relstorage = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_relstorages'), ',')) AND c.relkind = ANY (string_to_array(current_setting('arenadata_toolkit.tracking_relkinds'), ',')) ) SELECT bool_and(sc.cnt = pc.cnt) FROM segment_counts sc, pg_class_count pc;
 bool_and 
----------
 t        
(1 row)

1: SELECT arenadata_toolkit.tracking_unregister_db();
 tracking_unregister_db 
------------------------
 t                      
(1 row)
1: DROP EXTENSION arenadata_toolkit;
DROP

1q: ... <quitting>
2q: ... <quitting>

!\retcode gpconfig -r shared_preload_libraries;
(exited with code 0)
!\retcode gpconfig -r arenadata_toolkit.tracking_worker_naptime_sec;
(exited with code 0)
!\retcode gpstop -raq -M fast;
(exited with code 0)
